sdk/azcore/runtime/policy_api_version.go

```diff
@@ -0,0 +1,80 @@
+//go:build go1.18
+// +build go1.18
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package runtime
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+)
+
+const (
+	apiVersionHeaderName = iota
+	apiVersionQueryParamName
+)
+
+// APIVersionName is the name of an API version query parameter or header.
```

Comment:
I think you could simplify this a lot. How about this:

type APIVersionLocation int

const (
	APIVersionLocationQueryParam = iota
	APIVersionLocationHeader     = 1
)

// NewAPIVersionPolicy constructs an APIVersionPolicy. If version is "", Do will be a no-op.
func NewAPIVersionPolicy(location APIVersionLocation, name string, version string) *APIVersionPolicy {
	return &APIVersionPolicy{location: location, name: name, version: version}
}

// APIVersionPolicy enables users to set the API version of every request a client sends.
type APIVersionPolicy struct {
	// location indicages if 'name' refers to a query parameter or header.
	location APIVersionLocation

	// name of the query param or header whose value should be overridden; provided by the client.
	name string

	// version is the value (provided by the user) that replaces the default version value.
	version string
}

// Do sets the request's API version, if the policy is configured to do so, replacing any prior value.
func (a *APIVersionPolicy) Do(req *policy.Request) (*http.Response, error) {
	if a.version != "" {
		if a.name == "" {
			// user set ClientOptions.APIVersion but the client ctor didn't set PipelineOptions.APIVersionName
			return nil, errors.New("this client doesn't support overriding its API version")
		}
		switch a.location {
		case APIVersionLocationHeader:
			req.Raw().Header.Set(a.name, a.version)
		case APIVersionLocationQueryParam:
			q := req.Raw().URL.Query()
			q.Set(a.name, a.version)
			req.Raw().URL.RawQuery = q.Encode()
		}
	}
	return req.Next()
}

Reply:
I like it. The downside is that it `PipelineOptions` then needs two fields, one for the location and another for the name, and we won't be able to tell whether the caller set the location because its zero value will be useable. I suppose that's okay for an API intended for SDK developers.